### PR TITLE
Add EVP demo for X25519 key exchange

### LINKED ISSUE #14118: Add an EVP demo for key agreement
It should be the current best practice (fetch), have proper error handling and resource allocation/deallocation and be in the style of a known answer test.

### PR DESCRIPTION
**Pull request by hlandau**
This offers both a known answer test with fixed keys and also demonstrates a more realistic usage with random keys.

Fixes #14118.

### CODE REVIEW

**Comment by t8m** [comment]
Good work!

I'd suggest refactoring it slightly though so you do not do everything twice. Just split key generation (or key load) and the key exchange into separate functions and call these from main() for peer1 and peer2. This will make the demo even more real-world.

**Comment by hlandau** [comment]
@t8m How's this?

File: demos/keyexch/x25519.c [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+     * However for exposition, the code below demonstrates a generic
+     * implementation for arbitrary lengths.
+     */
+    if (local_peer->secret_len != 32) { /* unreachable */
+        fprintf(stderr, "Secret is always 32 bytes for X25519\n");
+        goto end;
+    }
+
+    /* Allocate memory for shared secrets. */
+    local_peer->secret = OPENSSL_zalloc(local_peer->secret_len);
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:170]: `OPENSSL_malloc` is probably sufficient here

File: demos/keyexch/x25519.c [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    if (local_peer->secret == NULL) {
+        fprintf(stderr, "Could not allocate memory for secret\n");
+        goto end;
+    }
+
+    /* Derive the shared secret. */
+    if (EVP_PKEY_derive(ctx, local_peer->secret,
+            &local_peer->secret_len) == 0) {
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:178]: I would line this up with `ctx`

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: demos/keyexch/x25519.c [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    const unsigned char *kat_privk_data,
+    PEER_DATA *local_peer)
+{
+    int rv = 0;
+    size_t pubk_data_len = 0;
+
+    /* Generate or load X25519 key for the peer */
+    if (kat_privk_data)
+        local_peer->privk = EVP_PKEY_new_raw_private_key_ex(libctx, "X25519", propq,
+            kat_privk_data, sizeof(peer1_privk_data));
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:74]: Sorry didn't spot this on the first run through...I'd split this like so:

```` C
        local_peer->privk =
            EVP_PKEY_new_raw_private_key_ex(libctx, "X25519", propq,
                                            kat_privk_data,
                                            sizeof(peer1_privk_data));
````

We generally try to line things up with the paren. The coding style is a bit vague on this though:

"Statements longer than 80 columns must be broken into sensible chunks, unless exceeding 80 columns significantly increases readability and does not hide information. Descendants are always substantially shorter than the parent and are placed substantially to the right."

There are several other places like this. I'd review all line breaks with this in mind.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: demos/keyexch/x25519.c [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+static int keyexch_x25519_before(
+    OSSL_LIB_CTX *libctx,
+    const unsigned char *kat_privk_data,
+    PEER_DATA *local_peer)
+{
+    int rv = 0;
+    size_t pubk_data_len = 0;
+
+    /* Generate or load X25519 key for the peer */
+    if (kat_privk_data)
```
Comment (t8m) [inline demos/keyexch/x25519.c:72]: Please always use explicit checks for NULL/nonNULL: `kat_privk_data != NULL`

File: demos/keyexch/x25519.c [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    if (local_peer->privk == NULL) {
+        fprintf(stderr, "Could not load or generate private key\n");
+        goto end;
+    }
+
+    /* Get public key corresponding to the private key */
+    if (EVP_PKEY_get_octet_string_param(local_peer->privk,
+            OSSL_PKEY_PARAM_PUB_KEY,
+            local_peer->pubk_data,
+            sizeof(local_peer->pubk_data), &pubk_data_len) == 0) {
```
Comment (t8m) [inline demos/keyexch/x25519.c:87]: Nit: Please align with the parenthesis.

File: demos/keyexch/x25519.c [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        fprintf(stderr, "EVP_PKEY_get_octet_string_param() failed\n");
+        goto end;
+    }
+
+    /* X25519 public keys are always 32 bytes */
+    if (pubk_data_len != 32) {
+        fprintf(stderr, "EVP_PKEY_get_octet_string_param() "
+            "yielded wrong length\n");
```
Comment (t8m) [inline demos/keyexch/x25519.c:95]: Nit: Align with the string above or with the parenthesis.

File: demos/keyexch/x25519.c [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+{
+    int rv = 0;
+    EVP_PKEY *remote_peer_pubk = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+
+    local_peer->secret = NULL;
+
+    /* Load public key for remote peer. */
+    remote_peer_pubk = EVP_PKEY_new_raw_public_key_ex(libctx, "X25519", propq,
+        remote_peer_pubk_data, 32);
```
Comment (t8m) [inline demos/keyexch/x25519.c:128]: Please align with parenthesis.

**Review by t8m — CHANGES_REQUESTED** [review]
Nice work! Just a few nits.

**Comment by hlandau** [comment]
Style fixes pushed.

File: demos/keyexch/x25519.c [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    0xb1, 0x76, 0xd7, 0x0e, 0x7e, 0x0f, 0xb6, 0xf4,
+    0x8c, 0x4a, 0x8c, 0x5f, 0xd8, 0x15, 0x09, 0x0a,
+    0x71, 0x78, 0x74, 0x92, 0x0f, 0x85, 0xc8, 0x43
+};
+
+static const unsigned char expected_result[32] = {
+  0x19, 0x71, 0x26, 0x12, 0x74, 0xb5, 0xb1, 0xce,
+  0x77, 0xd0, 0x79, 0x24, 0xb6, 0x0a, 0x5c, 0x72,
+  0x0c, 0xa6, 0x56, 0xc0, 0x11, 0xeb, 0x43, 0x11,
+  0x94, 0x3b, 0x01, 0x45, 0xca, 0x19, 0xfe, 0x09
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:47]: Wrong indentation here (4 spaces)

**Review by t8m — APPROVED** [review]

File: demos/keyexch/x25519.c [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    OSSL_LIB_CTX *libctx = NULL;
+    PEER_DATA peer1 = {"peer 1"}, peer2 = {"peer 2"};
+
+    /*
+     * Each peer generates its private key and sends its public key
+     * to the other peer. The private key is stored locally for
+     * later use.
+     */
+    if (keyexch_x25519_before(libctx, use_kat ? peer1_privk_data : NULL,
+            &peer1) == 0)
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:215]: Missed one

File: demos/keyexch/x25519.c [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        return 0;
+
+    if (keyexch_x25519_before(libctx, use_kat ? peer2_privk_data : NULL,
+            &peer2) == 0)
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:219]: and another

File: demos/keyexch/x25519.c [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+     * using memcmp(3).
+     */
+    if (CRYPTO_memcmp(peer1.secret, peer2.secret, peer1.secret_len) != 0) {
+        fprintf(stderr, "Negotiated secrets do not match\n");
+        goto end;
+    }
+
+    /* If we are doing the KAT, the secret should equal our reference result. */
+    if (use_kat && CRYPTO_memcmp(peer1.secret, expected_result,
+            peer1.secret_len) != 0) {
```
Comment (mattcaswell) [inline demos/keyexch/x25519.c:248]: and here

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Few more style nits...

**Comment by hlandau** [comment]
More style fixes.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to both branches.  Thanks.

**Pull request closed without merging**